Source/DafnyCore/Resolver.cs

```diff
@@ -4810,6 +4810,8 @@ public class ModuleBindings {
       Contract.Requires(sub != null && !(sub is TypeProxy));
       Contract.Requires(errorMsg != null);
       List<int> polarities = ConstrainTypeHead_Recursive(super, ref sub);
+      // Special case: strings.
+      super = super.AsSeqType is { } nonNull ? nonNull : super;
```

Comment:
Although this does patch the problem for strings, it is not the right general solution. For example, the following two test cases reveal that the problem is more general:

``` dafny
type IntSeq = seq<int>
method Test0()
{
  var a;
  var b := [a];
  var s: IntSeq := a;
}

type BoolSet = set<bool>
method Test1()
{
  var a;
  var b := {a};
  var s: BoolSet := a;
}
```

The general problem is that `super` is a type synonym that fixes some type parameter of the type its based on. This means that `super` and `sub` have different numbers of type arguments. To allow `super` and `sub` to be type synonyms, the enclosing method, `ImposeSubtypingConstraint`, should start by doing:

``` c#
      super = super.NormalizeExpandKeepConstraints();
      sub = sub.NormalizeExpandKeepConstraints();
```

(Apparently, `ConstrainTypeHead_Recursive` already does this and since `sub` is a `ref` parameter to it, `sub` will have been normalized-and-expanded-keeping-constrainted after the call to `ConstrainTypeHead_Recursive`. Still, I suggest doing this for both `super` and `sub` in `ImposeSubtypingConstraint` (as I showed above), since that seems both more robust and less surprising.)

I also suggest adding Clément's two small test cases from #2829 as well as my two test cases from above. You may also add

``` dafny
datatype Option<X> = None | Some(X)
type OptionReal = Option<real>
method Test2()
{
  var a;
  var b := Some(a);
  var s: OptionReal := a;
}
```

even though this test apparently already works (I didn't look to see where, but it seems that `NormalizeExpandKeepConstraints()` has already been applied to `super` in this case).